Add RRG

Srujan-D opened the pull request.

Srujan-D committed:
Create rrg.py

Srujan-D committed:
Create rrg_test.py

threewisemonkeys-as commented:
Could you chnage the API to the new use samplers now? You can check out the curent sampling based algos for examples

Srujan-D commented:
Yeah, i'll do that. 
Also, A-Star is taking a lot of time (like, more than 30 minutes at least) to run. So, either there's a bug in rrg (mostly, but when I plot the graph at the end, I get this
![Screenshot from 2020-07-30 16-39-40](https://user-images.githubusercontent.com/53463585/88922402-b2c99b00-d28d-11ea-9591-04a78c0dd32a.png)
) or i need another algorithm for searching the shortest path (maybe dijkstra?)
Also, that image was for 1 obstacle only. I need help in making the code work for multiple obstacles, like my graph itself should have a graph which works fine with all the obstacles.

Srujan-D committed:
Update rrg_test.py

Srujan-D committed:
Update rrg.py

Srujan-D committed:
fixed lint errors.

hardesh commented:
@Srujan-D A* shouldn't take this much time. In fact, it should run faster than RRG. I think there's a bug in your RRG.

Srujan-D commented:
@hardesh Yeah, that's what I am searching for. My RRG actually doesn't take much time to run, but for 500 (max)iterations in RRG, the `len(open_)` from the A* code goes up to approx. 8000. Also, this length is not increasing/decreasing linearly, it goes up and down many many times.That's where (the while loop) majority of the time goes.

Srujan-D committed:
Update rrg_test.py

Srujan-D committed:
Update rrg.py

Srujan-D committed:
Fixed LGTM Alerts.

Srujan-D commented:
Sort? 
Also, should I make those changes we discussed (in the meet) in the graph class in this PR itself?

Srujan-D commented:
Also, i think there's some problem with the get_status method in envs. It was causing the RRT bug which got sorted by removing get_status from the code, and the same happened with RRG.

threewisemonkeys-as commented:
Is rrg + astar working properly now?

Srujan-D committed:
Add rrg.

Srujan-D commented:
> Is rrg + astar working properly now?

Yep, there were 2-3 problems in the rrg code which all contributed to this. I am able to see the path properly now. So it's all good now.

threewisemonkeys-as commented on gennav/planners/rrt/rrg.py:
```diff
@@ -0,0 +1,124 @@
+
+        Args:
+
+            sampler (function): samples random points in sample_area
```
```suggestion
            sampler (gennav.utils.sampler.Sampler): sampler to get random states
```

threewisemonkeys-as commented on gennav/planners/rrt/rrg.py:
```diff
@@ -0,0 +1,124 @@
+        Args:
+
+            sampler (function): samples random points in sample_area
+            goal_sample_p (float, optional): Probability of sampling goal. Defaults to 0.1
```
```suggestion
```

threewisemonkeys-as commented on gennav/planners/rrt/rrg.py:
```diff
@@ -0,0 +1,124 @@
+    def plan(self, start, goal, env):
+        """Plans path from start to goal avoiding obstacles.
+        Args:
+            start_point(gennav.utils.RobotState): with start point coordinates.
+            end_point (gennav.utils.RobotState): with end point coordinates.
```
```suggestion
            start (gennav.utils.RobotState): Starting state
            end (gennav.utils.RobotState): Goal state
```

threewisemonkeys-as commented on tests/test_planners/rrg_test.py:
```diff
@@ -0,0 +1,31 @@
+    for obstacles in general_obstacles_list:
+        poly.update(obstacles)
+
+        start = RobotState(position=Point(0, 0))
+        goal = RobotState(position=Point(12, 10))
+        my_tree = RRG(sampler=sampler, expand_dis=1.0, max_iter=500)
```
These lines can be taken out the loop. You can check out the other tests as examples.

Srujan-D replied:
Line 23 and 24 can be taken out of the loop. If I take the 25th line out, then I am getting the same path for both the obstacles, which leads to an assertion error for the 2nd obstacle. This is because my `self.graph = Graph()` is in `__init__` . So I'll have to remove that and write `graph = Graph()` in `plan` itself. I'll update it shortly.

Srujan-D committed:
Update rrg.py

1) Updated doc string
2) Changed 'goal' to 'end'
3) Initialized graph in plan itself.

Srujan-D committed:
Removed unnecessary lines from the obstacles loop.

Srujan-D commented:
Should I update the graph class? (Replacing from_node and to_node by node1 and node2 respectively)

threewisemonkeys-as commented:
> Should I update the graph class? (Replacing from_node and to_node by node1 and node2 respectively)

Yeah sure

Srujan-D committed:
Update graph.py

1) Changed 'from_node' to 'node1'.
2) Changed 'to_node' to 'node2'.

Srujan-D commented:
If no other changes are required, I think we can merge this.

threewisemonkeys-as commented:
Resolves #6

threewisemonkeys-as commented:
Great work!
